sdk/core/core/src/az_http_policy.c

```diff
@@ -61,12 +65,110 @@ AZ_NODISCARD az_result az_http_pipeline_policy_retry(
     void * const data,
     az_http_request_builder * const hrb,
     az_http_response * const response) {
-  (void)data;
-
   // reset response to be written from the start
-  AZ_RETURN_IF_FAILED(az_http_response_reset(response));
+  if (data == NULL) {
+    AZ_RETURN_IF_FAILED(az_http_response_reset(response));
+    return az_http_pipeline_nextpolicy(p_policies, hrb, response);
+  }
 
-  return az_http_pipeline_nextpolicy(p_policies, hrb, response);
+  az_http_policy_retry_options const * const retry_policy
+      = (az_http_policy_retry_options const *)data;
+
+  int16_t const max_tries = retry_policy->max_tries;
+
+  bool const custom_retry_delay_limits = retry_policy->retry_delay_msec > 0
+      && retry_policy->max_retry_delay_msec > 0
+      && (retry_policy->retry_delay_msec <= retry_policy->max_retry_delay_msec);
+
+  int32_t const retry_delay_msec = custom_retry_delay_limits
+      ? retry_policy->retry_delay_msec
+      : _az_HTTP_POLICY_RETRY_OPTIONS_DEFAULT_RETRY_DELAY_MSEC;
+
+  int32_t const max_retry_delay_msec = custom_retry_delay_limits
+      ? retry_policy->max_retry_delay_msec
+      : _az_HTTP_POLICY_RETRY_OPTIONS_DEFAULT_MAX_RETRY_DELAY_MSEC;
+
+  bool const should_log = az_log_should_write(AZ_LOG_HTTP_RETRY);
+  az_result result = AZ_OK;
+  int16_t attempt = 1;
+  while (true) {
+    AZ_RETURN_IF_FAILED(az_http_response_reset(response));
+    result = az_http_pipeline_nextpolicy(p_policies, hrb, response);
+
+    // Even HTTP 429, or 502 are expected to be AZ_OK, so the failed result is not retriable.
+    if (attempt > max_tries || az_failed(result) || response == NULL
+        || response->builder.length == 0) {
+      return result;
+    }
+
+    int32_t retry_after_msec = -1;
+    {
+      az_http_response_parser response_parser = { 0 };
+      AZ_RETURN_IF_FAILED(az_http_response_parser_init(
+          &response_parser, az_span_builder_result(&response->builder)));
+
+      az_http_response_status_line status_line = { 0 };
+      AZ_RETURN_IF_FAILED(az_http_response_parser_read_status_line(&response_parser, &status_line));
+
+      switch (status_line.status_code) {
+        // Keep retrying on the HTTP response codes below, return immediately otherwise.
+        case 408:
+        case 429:
+        case 500:
+        case 502:
+        case 503:
+        case 504: {
+          // Try to get the value of retry-after header, if there's one.
+          az_pair header = { 0 };
+          while (az_http_response_parser_read_header(&response_parser, &header) == AZ_OK) {
+            if (az_span_is_equal(header.key, AZ_STR("retry-after-ms"))
```

Comment:
These comparisons need to be case-insensitive. I suggest putting the code to find the headers in a separate private function to make the retry logic cleaner. This code tends to have bugs and be hard to follow. Breaking it into discreet chunks helps with code maintainability a lot.